CactbotOverlay/FFXIVProcess.cs

```diff
@@ -570,522 +486,301 @@ public class EntityData {
         // The pointer can be null when not logged in.
         return null;
       }
-
+      job_inner_ptr = IntPtr.Add(job_inner_ptr, kJobDataInnerStructOffset);
       return Read8(job_inner_ptr, kJobDataInnerStructSize);
     }
 
-    public class RedMageJobData {
-      public int white = 0;
-      public int black = 0;
-
-      public override bool Equals(object obj) {
-        var o = obj as RedMageJobData;
-        return o != null &&
-          white == o.white &&
-          black == o.black;
-      }
-
-      public override int GetHashCode() {
-        int hash = 17;
-        hash = hash * 31 + white.GetHashCode();
-        hash = hash * 31 + black.GetHashCode();
-        return hash;
-      }
-    }
-
-    public RedMageJobData GetRedMage() {
-      byte[] bytes = GetJobSpecificData();
-      if (bytes == null)
+    public unsafe JObject GetJobSpecificData(EntityJob job) {
+      if (!HasProcess() || job_data_outer_addr_ == IntPtr.Zero)
         return null;
 
-      var j = new RedMageJobData();
-      j.white = bytes[kJobDataInnerStructOffsetJobSpecificData];
-      j.black = bytes[kJobDataInnerStructOffsetJobSpecificData + 1];
-      return j;
-    }
-
-    public class WarriorJobData {
-      public int beast = 0;
-
-      public override bool Equals(object obj) {
-        var o = obj as WarriorJobData;
-        return o != null &&
-          beast == o.beast;
-      }
-
-      public override int GetHashCode() {
-        int hash = 17;
-        hash = hash * 31 + beast.GetHashCode();
-        return hash;
-      }
-    }
-
-    public WarriorJobData GetWarrior() {
-      byte[] bytes = GetJobSpecificData();
-      if (bytes == null)
+      IntPtr job_inner_ptr = ReadIntPtr(job_data_outer_addr_);
+      if (job_inner_ptr == IntPtr.Zero) {
+        // The pointer can be null when not logged in.
         return null;
-
-      var j = new WarriorJobData();
-      j.beast = bytes[kJobDataInnerStructOffsetJobSpecificData];
-      return j;
-    }
-
-    public class DarkKnightJobData {
-      public int blood = 0;
-      public int darkside_ms = 0;
-
-      public override bool Equals(object obj) {
-        var o = obj as DarkKnightJobData;
-        return o != null &&
-          blood == o.blood &&
-          darkside_ms == o.darkside_ms;
       }
-
-      public override int GetHashCode() {
-        int hash = 17;
-        hash = hash * 31 + blood.GetHashCode();
-        hash = hash * 31 + darkside_ms.GetHashCode();
-        return hash;
+      job_inner_ptr = IntPtr.Add(job_inner_ptr, kJobDataInnerStructOffset);
+
+      fixed (byte* p = Read8(job_inner_ptr, kJobDataInnerStructSize)) {
+        if (p == null)
+          return null;
+        else {
+          switch (job) {
+            case EntityJob.RDM: {
+                return JObject.FromObject(*(RedMageJobMemory*)&p[0]);
```

Comment:
What I wonder is if RedMageJobMemory could just be serialized itself, rather than going job memory -> jobject -> job detail classes.  Just, job memory, that's it.

You can add properties that are serialized, like this: https://github.com/quisquous/ACT_EnmityPlugin/blob/master/EnmityPlugin/FFXIVObject.cs#L51

You can also disallow properties to be serialized with `[NonSerialized]`.

Between the two of those, I think you should be able to make it so that the "job memory" classes serialize to the same as the "job detail" classes.  In most cases, they should be identical?

Reply:
Ahhh that's embarrassingly simple... I can make those changes and strip out all the JobDetail classes from JSEvents.

Reply:
I've pushed a new build with this implemented for you to review when you can :)